tools/executable/executable.go

```diff
@@ -0,0 +1,86 @@
+package executable
+
+import (
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/errors"
+)
+
+type ExecutableMCMSProposalBase struct {
+	Version              string      `json:"version"`
+	ValidUntil           uint32      `json:"validUntil"`
+	Signatures           []Signature `json:"signatures"`
+	OverridePreviousRoot bool        `json:"overridePreviousRoot"`
+
+	// Map of chain identifier to chain metadata
+	ChainMetadata map[string]ExecutableMCMSChainMetadata `json:"chainMetadata"`
+}
+
+type ExecutableMCMSChainMetadata struct {
+	NonceOffset uint64         `json:"nonceOffset"`
+	MCMAddress  common.Address `json:"mcmAddress"`
+}
+
+func (m ExecutableMCMSProposalBase) Validate() error {
+	if m.Version == "" {
+		return &errors.ErrInvalidVersion{
+			ReceivedVersion: m.Version,
+		}
+	}
+
+	// Get the current Unix timestamp as an int64
+	currentTime := time.Now().Unix()
+
+	if m.ValidUntil <= uint32(currentTime) {
+		// ValidUntil is a Unix timestamp, so it should be greater than the current time
+		return &errors.ErrInvalidValidUntil{
+			ReceivedValidUntil: m.ValidUntil,
+		}
+	}
+
+	if len(m.ChainMetadata) == 0 {
+		return &errors.ErrNoChainMetadata{}
+	}
+
+	return nil
+}
+
+type ExecutableMCMSProposal struct {
+	ExecutableMCMSProposalBase
+
+	// Operations to be executed
+	Transactions []ChainOperation `json:"transactions"`
+}
+
+func (m *ExecutableMCMSProposal) Validate() error {
+	if err := m.ExecutableMCMSProposalBase.Validate(); err != nil {
+		return err
+	}
+
+	if len(m.Transactions) == 0 {
+		return &errors.ErrNoTransactions{}
+	}
+
+	// Validate all chains in transactions have an entry in chain metadata
+	for _, t := range m.Transactions {
+		if _, ok := m.ChainMetadata[t.ChainIdentifier]; !ok {
+			return &errors.ErrMissingChainDetails{
+				ChainIdentifier: t.ChainIdentifier,
+				Parameter:       "chain metadata",
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m *ExecutableMCMSProposal) ToExecutor(clients map[string]ContractDeployBackend) (*Executor, error) {
```

Comment:
This factory sort of pattern would allow for a caller to create multiple executors for the same proposal which we would not want right? I think it makes more sense to have a executor object instantiated with clients which has a function like 
```
func (e Executor) ExecuteProposal(ep ExecutableMCMSProposal) error {}
```

Reply:
Executor isn't just used for execution though, it also holds the fully constructed MerkleTree which is used for signing as well though, so having a single `Execute` function wouldn't suffice. Executor just represents the Full expanded Proposal with all machine executable details required for signing, validating, and executing